### DISCUSSION TITLE
Using a PP value from previous timestep

### DISCUSSION THREAD

**Discussion by hokonkim** [idaholab/moose#19449] [Q&A General] [answered]
Hello, I am looking for a way to use a PP value from previous timestep.

V_Cr_last is a zero at first, and it should be replaced by computed value.

Based on previous posts, I tried several ways..but, I still cannot find an appropriate way to do this.

![image](https://user-images.githubusercontent.com/46009457/142833032-05313c6b-e007-491e-b982-41f44e3e9ae6.png)

    [Materials]
      [./CrDepoActivity]
        type  = ParsedMaterial
        block = 'PT_TPB_TET4'
        f_name = 'CrDepoActivity_SD'
        constant_names        = 'R          T        z     F          x_CrO2OH2   x_H2O  V_molar_Cr2O3'
                            # (J/K/mol),  (K), (# of e-), (C/mol),    (non),      (non), (cm^3)
        constant_expressions  = '8.3144598  1073.0   4   96485.3329   1e-8        0.01   29.12'
        function = '(SD_init * x_CrO2OH2^(0.5) * x_H2O^(0.5) * 2 * sinh(F * 2 / R / T * aux_eta_tpb)) *
                     V_molar_Cr2O3 / z / F'
        args = 'SD_init aux_eta_tpb'
        # postprocessor_names = ''
        # Make values see at output(*.e) file
        # outputs = exodus
      [../]
    []

    [AuxVariables]
      [./aux_eta_tpb]
        block = 'PT_TPB_TET4'
      [../]

      [./SD_init]
        block = 'PT_TPB_TET4'
        initial_condition = 1
      [../]
    []

    [Postprocessors]
      [./volume_TPB]
        type = VolumePostprocessor
        block = 'PT_TPB_TET4'
        outputs = 'console csv'
      [../]

      [./avg_SD_Cr2O3]
        type = ElementAverageMaterialProperty
        mat_prop = CrDepoActivity_SD
        block = 'PT_TPB_TET4'
        outputs = 'console csv'
      [../]

      [./V_Cr_last]
        type = ParsedPostprocessor
        function = 'avg_SD_Cr2O3 * (volume_TPB - 0) * Timestep_now'
        pp_names = 'avg_SD_Cr2O3 volume_TPB Timestep_now'
        outputs = 'console csv'
        # execute_on = 'initial timestep_end' # Need this line, because Postprocessor is not being executed at the 0th timestep.
      [../]

      [./dV_Cr_dt]
        type = ParsedPostprocessor
        function = 'avg_SD_Cr2O3 * (volume_TPB - V_Cr_last)'
        pp_names = 'avg_SD_Cr2O3 volume_TPB V_Cr_last'
        outputs = 'console csv'
      [../]

      [./V_Cr_new]
        type = ParsedPostprocessor
        function = 'V_Cr_last + (dV_Cr_dt * Timestep_now)'
        pp_names = 'V_Cr_last dV_Cr_dt Timestep_now'
        outputs = 'console csv'
        execute_on = 'initial timestep_end' # Need this line, because Postprocessor is not being executed at the 0th timestep.
      [../]

      [./Time_now]
        type = TimePostprocessor
        outputs = 'console csv'
      [../]

      [./Timestep_now]
        # delta time
        type = TimestepSize
        outputs = 'console csv'
      [../]
    []

**Comment by GiudGiud** [comment, marked as answer]
ParsedMaterial wont let you do that directly. 
You can either :
- code your own material and use getPostprocessorValueOld()
- hack something with changeOverTimePostprocessor (old = new - change)
- add an OldPostprocessor to the framework that uses getPostprocessorValueOld()

**Reply by hokonkim** [reply]
It looks like I need to make a custom postprocessor.

Based on changeOverTimePostprocessor and EmptyPostprocessor,

I made (custom) OldPostprocessor.c and .h files like below

This is the header file (OldPostprocessor.h).
```
    #pragma once

    #include "GeneralPostprocessor.h"

    // Forward Declarations
    class OldPostprocessor;

    template <>
    InputParameters validParams<OldPostprocessor>();

    class OldPostprocessor : public GeneralPostprocessor
    {
    public:
      static InputParameters validParams();

      OldPostprocessor(const InputParameters & parameters);

      virtual void initialize() override;
      virtual void execute() override;
      virtual Real getValue() override;

    protected:
      /// current post-processor value
      const PostprocessorValue & _pps_value;

      /// old post-processor value
      const PostprocessorValue & _pps_value_old;
    };


This is the source code (OldPostprocessor.C).


    #include "OldPostprocessor.h"

    #include "libmesh/parallel.h"

    registerMooseObject("ermineApp", OldPostprocessor);

    defineLegacyParams(OldPostprocessor);

    InputParameters
    OldPostprocessor::validParams()
    {
      InputParameters params = GeneralPostprocessor::validParams();
      params.addClassDescription("A postprocessor object that returns a value of zero.");
      params.addRequiredParam<PostprocessorName>("postprocessor", "The name of the postprocessor");
      return params;
    }

    OldPostprocessor::OldPostprocessor(const InputParameters & parameters)
      : GeneralPostprocessor(parameters),
      _pps_value(getPostprocessorValue("postprocessor")),
      _pps_value_old(getPostprocessorValueOld("postprocessor"))
    {
    }

    Real
    OldPostprocessor::getValue()
    {
      if(_t_step == 0)
      {
        return 0;
      }
      else
      {
        return _pps_value_old;
      }
    }
```

Did I do write a correct OldPostprocessor to get a previous time-step value of a Postprocessor?

**Reply by GiudGiud** [reply]
yes
If you add a boolean to switch between new, old and older I think we would welcome the contribution the framework

**Reply by hokonkim** [reply]
![image](https://user-images.githubusercontent.com/46009457/142945205-c5a3b9b6-8d3b-4cf0-9cb7-af7f732f56d6.png)

At a first step(initial value), V_Cr,prev should be put as a zero into d_Cr/dt and V_Cr,new.

But, in present input file and OldPostprocessor, I cannot do that.

Should I modify the initialize section in the source code?


    #include "OldPostprocessor.h"

    registerMooseObject("ermineApp", OldPostprocessor);

    defineLegacyParams(OldPostprocessor);

    InputParameters
    OldPostprocessor::validParams()
    {
      InputParameters params = GeneralPostprocessor::validParams();

      params.addRequiredParam<PostprocessorName>("postprocessor", "The name of the postprocessor");
      params.addClassDescription("A postprocessor that gets a postprocessor value from a previous time step");
      return params;
    }

    OldPostprocessor::OldPostprocessor(const InputParameters & parameters)
      : GeneralPostprocessor(parameters),
        _pps_value(getPostprocessorValue("postprocessor")),
        _pps_value_old(getPostprocessorValueOld("postprocessor"))
    {
    }

    void
    OldPostprocessor::initialize()
    {
        Real zeroVal = 0;   ??????????????????????
    }

    void
    OldPostprocessor::execute()
    {
    }

    Real
    OldPostprocessor::getValue()
    {
      if(_t_step == 0)
      {
        return zeroVal;
      }
      else
      {
        return _pps_value_old;
      }
    }


By the way, how can I do "add a boolean to switch between new, old and older"???

I have no idea about this..

**Reply by GiudGiud** [reply]
VCr_prev is not 0?

the boolean I mentioned would be an additional parameter. Could be a boolean or could be an enum. If you dont see what i m talking about just make it work for your case

**Reply by hokonkim** [reply]
So far, this codes didn't work for me...

If I include V_Cr_prev into dV_Cr_dt or V_Cr_new, then it didn't work.

I think referring values from itself is problem?

      [./dV_Cr_dt]
        type = ParsedPostprocessor
        function = 'if(Time_now = 0, avg_SD_Cr2O3 * (volume_TPB - 0), avg_SD_Cr2O3 * (volume_TPB - V_Cr_prev))'
        pp_names = 'Time_now avg_SD_Cr2O3 volume_TPB V_Cr_prev'
        outputs = 'console csv'
        # execute_on = 'initial timestep_end' # Need this line, because Postprocessor is not being executed at the 0th timestep.
      [../]

      [./V_Cr_new]
        type = ParsedPostprocessor
        function = 'if(Time_now = 0, 0, V_Cr_prev + (dV_Cr_dt * Timestep_now))'
        pp_names = 'Time_now V_Cr_prev dV_Cr_dt Timestep_now'
        outputs = 'console csv'
        execute_on = 'initial timestep_end' # Need this line, because Postprocessor is not being executed at the 0th timestep.
      [../]

      [./V_Cr_prev]
        type = OldPostprocessor
        postprocessor = V_Cr_new
        outputs = 'console csv'
        # execute_on = 'initial timestep_end' # Need this line, because Postprocessor is not being executed at the 0th timestep.
      [../]

      [./Time_now]
        type = TimePostprocessor
        outputs = 'console csv'
      [../]

      [./Timestep_now]
        # delta time
        type = TimestepSize
        outputs = 'console csv'
      [../]

**Reply by GiudGiud** [reply]
why does it not work for you?

**Reply by GiudGiud** [reply]
your getValue returns a 0 at time step 0, so VCr_prev should be 0 then. 
Same for VCr_new, it returns 0 at time 0. 

Which postprocessor is not equal to what you want?

**Reply by hokonkim** [reply]
<img width="686" alt="Screen Shot 2021-11-22 at 8 19 16 PM" src="https://user-images.githubusercontent.com/46009457/142958194-f915b79d-873f-42ad-b40c-9fe31d6ec3db.png">

I cannot run a simulation like above. It just terminated automatically.

So, I cannot check the values from each postprocessor after I modified the input file like previous reply.

**Reply by GiudGiud** [reply]
Ah that s different. Something is wrong in your postprocessor and it is crashing. 
You can try to compile in debug mode and find the problem. 

I had a second look at your code. 
What is zeroVal ? Is it defined in your header? What do you need that? It s just 0, you can just have return 0; in getValue for t=0

if it s a class attribute, then doing Real zeroVal = 0 in initalize will not write to it. Declaring the type like this means it ll create a new one (which will be released as soon as the initialize is done)

**Reply by GiudGiud** [reply]
also no more screenshots lol 
please paste in quotes

**Reply by hokonkim** [reply]
This is the present source code for OldPostprocessor.

    #include "OldPostprocessor.h"

    registerMooseObject("ermineApp", OldPostprocessor);

    defineLegacyParams(OldPostprocessor);

    InputParameters
    OldPostprocessor::validParams()
    {
      InputParameters params = GeneralPostprocessor::validParams();

      params.addRequiredParam<PostprocessorName>("postprocessor", "The name of the postprocessor");
      params.addClassDescription("A postprocessor that gets a postprocessor value from a previous time step");
      return params;
    }

    OldPostprocessor::OldPostprocessor(const InputParameters & parameters)
      : GeneralPostprocessor(parameters),
        _pps_value(getPostprocessorValue("postprocessor")),
        _pps_value_old(getPostprocessorValueOld("postprocessor")),
        _pps_value_initial(declareRestartableData<Real>("pps_value_initial"))
    {
    }

    void
    OldPostprocessor::initialize()
    {
      _pps_value_initial = 0;
    }

    void
    OldPostprocessor::execute()
    {
    }

    Real
    OldPostprocessor::getValue()
    {
      if(_t_step == 0)
      {
        return _pps_value_initial;
      }
      else
      {
        return _pps_value_old;
      }
    }



Like you said, something was crashed if I put OldPostprocessor into other PPs.

**Reply by hokonkim** [reply]
In debug mode, I got messages like below

    Process 51507 stopped
    * thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
        frame #0: 0x0000000113292120 libmpi.12.dylib`MPI_Abort
    libmpi.12.dylib`MPI_Abort:
    ->  0x113292120 <+0>: pushq  %rbp
        0x113292121 <+1>: movq   %rsp, %rbp
        0x113292124 <+4>: pushq  %r15
        0x113292126 <+6>: pushq  %r14
    Target 0: (ermine-dbg) stopped.


(typing bt)

> * thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
>   * frame #0: 0x0000000113292120 libmpi.12.dylib`MPI_Abort
>     frame #1: 0x000000010868b718 libmesh_dbg.0.dylib`libMesh::libmesh_terminate_handler() + 72
>     frame #2: 0x00007fff516307c9 libc++abi.dylib`std::__terminate(void (*)()) + 8
>     frame #3: 0x00007fff5163026f libc++abi.dylib`__cxa_throw + 121
>     frame #4: 0x0000000101eae4d0 libmoose-dbg.0.dylib`DependencyResolver<DependencyResolverInterface*>::insertDependency(this=0x00007ffeefbfc518, key=0x000000013cc12b20, value=0x000000013cc12b18) at DependencyResolver.h:237
>     frame #5: 0x0000000101ea860c libmoose-dbg.0.dylib`void DependencyResolverInterface::sort<DependencyResolverInterface*>(vector=size=7) at DependencyResolverInterface.h:84
>     frame #6: 0x0000000101ea0368 libmoose-dbg.0.dylib`TheWarehouse::prepare(this=0x0000000113b704f0, conds=<unavailable>) at TheWarehouse.C:174
>     frame #7: 0x0000000101e9fbfe libmoose-dbg.0.dylib`TheWarehouse::queryID(this=0x0000000113b704f0, conds=size=4) at TheWarehouse.C:211
>     frame #8: 0x0000000100e5bbb0 libmoose-dbg.0.dylib`std::__1::vector<GeneralUserObject*, std::__1::allocator<GeneralUserObject*> >& TheWarehouse::QueryCache<>::queryInto<GeneralUserObject>(this=0x00007ffeefbfdc08, results=size=0) at TheWarehouse.h:271
>     frame #9: 0x0000000100e59ca7 libmoose-dbg.0.dylib`FEProblemBase::computeUserObjectsInternal(this=0x0000000115028418, type=0x000000010234e0b0, group=0x00007ffeefbfde20, query=0x00007ffeefbfdd68) at FEProblemBase.C:3858
>     frame #10: 0x0000000100e599e0 libmoose-dbg.0.dylib`FEProblemBase::computeUserObjects(this=0x0000000115028418, type=0x000000010234e0b0, group=0x00007ffeefbfde20) at FEProblemBase.C:3842
>     frame #11: 0x0000000100e5862f libmoose-dbg.0.dylib`FEProblemBase::execute(this=0x0000000115028418, exec_type=0x000000010234e0b0) at FEProblemBase.C:3755
>     frame #12: 0x00000001007ab4c9 libmoose-dbg.0.dylib`FixedPointSolve::solveStep(this=0x000000013b5789d0, begin_norm=0x000000013aa542b0, end_norm=0x000000013ab0c630, transformed_dofs=size=0) at FixedPointSolve.C:419
>     frame #13: 0x00000001007a9061 libmoose-dbg.0.dylib`FixedPointSolve::solve(this=0x000000013b5789d0) at FixedPointSolve.C:266
>     frame #14: 0x000000010154dae0 libmoose-dbg.0.dylib`TimeStepper::step(this=0x000000013b5cd938) at TimeStepper.C:163
>     frame #15: 0x00000001007b0f74 libmoose-dbg.0.dylib`Transient::takeStep(this=0x000000011480e418, input_dt=-1) at Transient.C:437
>     frame #16: 0x00000001007bd3fb libmoose-dbg.0.dylib`Transient::execute(this=0x000000011480e418) at Transient.C:322
>     frame #17: 0x0000000101f5fc50 libmoose-dbg.0.dylib`MooseApp::executeExecutioner(this=0x000000011404bc18) at MooseApp.C:1061
>     frame #18: 0x0000000101f59b48 libmoose-dbg.0.dylib`MooseApp::run(this=0x000000011404bc18) at MooseApp.C:1289
>     frame #19: 0x000000010000cfb5 ermine-dbg`main(argc=3, argv=0x00007ffeefbffb78) at main.C:33
>     frame #20: 0x00007fff53671015 libdyld.dylib`start + 1
>     frame #21: 0x00007fff53671015 libdyld.dylib`start + 1

**Reply by GiudGiud** [reply]
oh interesting the dependency resolver is dying on this.

**Reply by hokonkim** [reply]
How can I handle this issue? I really have no idea....

**Reply by GiudGiud** [reply]
oh here s your problem
```
  [./V_Cr_new]
    type = ParsedPostprocessor
    function = 'if(Time_now = 0, 0, V_Cr_prev + (dV_Cr_dt * Timestep_now))'
    pp_names = 'Time_now V_Cr_prev dV_Cr_dt Timestep_now'
    outputs = 'console csv'
    execute_on = 'initial timestep_end' # Need this line, because Postprocessor is not being executed at the 0th timestep.
  [../]

  [./V_Cr_prev]
    type = OldPostprocessor
    postprocessor = V_Cr_new
    outputs = 'console csv'
    # execute_on = 'initial timestep_end' # Need this line, because Postprocessor is not being executed at the 0th timestep.
  [../]
```

V_Cr_new depends on V_Cr_prev, which depends on V_Cr_new.

You cant do that, the system will not know which one to execute first. So you have to find a way to break that chain.
An easy one is to just make your own postprocessor for V_Cr_new


OR

you could use a TimeIntegratedPostprocessor on dVCr_dt to compute V_Cr_new ??
https://mooseframework.inl.gov/source/postprocessors/TimeIntegratedPostprocessor.html
